ledger/acctupdates.go

```diff
@@ -776,10 +777,94 @@ func accountHashBuilder(addr basics.Address, accountData basics.AccountData, enc
 }
 
 // Initialize accounts DB if needed and return account round
-func (au *accountUpdates) accountsInitialize(tx *sql.Tx) (basics.Round, error) {
-	err := accountsInit(tx, au.initAccounts, au.initProto)
+func (au *accountUpdates) accountsInitialize(ctx context.Context, tx *sql.Tx) (basics.Round, error) {
+	// check current daabase version.
+	dbVersion, err := db.GetUserVersion(ctx, tx)
 	if err != nil {
-		return 0, err
+		return 0, fmt.Errorf("accountsInitialize unable to read database schema version : %v", err)
+	}
+
+	// if database version is greater than supported by current binary, write a warning. This would keep the existing
+	// fallback behaviour where we could use an older binary iff the schema happen to be backward compatible.
+	if dbVersion > accountDBVersion {
+		au.log.Warnf("accountsInitialize database schema version is %d, but algod supports only %d", dbVersion, accountDBVersion)
+	}
+
+	if dbVersion < accountDBVersion {
+		au.log.Infof("accountsInitialize upgrading database schema from version %d to version %d", dbVersion, accountDBVersion)
+
+		for dbVersion < accountDBVersion {
+			au.log.Infof("accountsInitialize performing upgrade from version %d", dbVersion)
+			// perform the initialization/upgrade
+			switch dbVersion {
```

Comment:
this could get to be a huge blob of code after a couple more versions; might be worth setting a pattern now of putting each upgrade in a member function `func (au *accountUpdates) dbUpgrade1(ctx,tx)(error){}` or so.